### ISSUE TITLE
Race condition: Goal state is SUCCEEDED before result is available.

### ISSUE THREAD

**Issue by jorgenfb** [ros/actionlib#116]
SimpleActionClient seems to have a race condition. Calling `get_state()` can give the SUCCEEDED state before a result is available.

This is a package showing the problem. It uses a `SimpleActionClient` to make a request and then checks `client.get_state()` at a high frequency to check for `SUCCEEDED` state and then checks if `client.get_result()` returns a result. An error is printed if the result is missing. It makes a maximum of 1000 requests to try to trigger this error. Run `roslaunch action_client_breaker action_client_breaker.test` and wait for a ERROR message in the console.
[action_client_breaker.zip](https://github.com/ros/actionlib/files/2522118/action_client_breaker.zip)

BufferClient in tf2_ros was previously affected by this issue. See ros/geometry2#178.

**Comment by HappySamuel** [comment]
Hi @jorgenfb 

I also faced this issue when i am calling client.get_state() at around 5Hz. Have you find a solution / workaround for this?

Best,
Samuel

**Comment by jorgenfb** [comment]
Yes, have a look at https://github.com/ros/geometry2/issues/178. Especially this comment: https://github.com/ros/geometry2/issues/178#issuecomment-431801344. You might be able to use:
```
your_simple_action_client.wait_for_result()
```

**Comment by HappySamuel** [comment]
Isn't the `client.wait_for_result()` will put whole system on wait? until result come out?

**Comment by jorgenfb** [comment]
Just your current thread. If you are only spinning until a result is ready,
then you can use wait_for_result without any issue
Regards
Jørgen Borgesen
nLink
+47 456 19 649

The content of this e-mail is of a confidential and proprietary nature and
may be protected by law. The content is only intended for the individual(s)
named in the message. If you have received this e-mail by mistake, please
reply to this message and delete the e-mail and its content afterwards. You
may not copy or distribute any of the content in this e-mail.

**Comment by sloretz** [comment]
ROS 1 is end-of-life (EOL) as of today, May 31st 2025. I am archiving this repository because:

* it only supports ROS 1
* it isn't needed anymore in ROS 2

[If you still rely on ROS 1, read this page to learn about your options.](https://bit.ly/NoeticEOL)
